include/swift/AST/Types.h

```diff
@@ -2844,39 +2863,62 @@ class AnyFunctionType : public TypeBase {
     // If bits are added or removed, then TypeBase::AnyFunctionTypeBits
     // and NumMaskBits must be updated, and they must match.
     //
-    //   |representation|noEscape|throws|
-    //   |    0 .. 3    |    4   |   5  |
+    //   |representation|noEscape|throws|uncommon|
+    //   |    0 .. 3    |    4   |   5  |    6   |
     //
     enum : unsigned {
       RepresentationMask     = 0xF << 0,
       NoEscapeMask           = 1 << 4,
       ThrowsMask             = 1 << 5,
-      NumMaskBits            = 6
+      UncommonMask           = 1 << 6,
+      NumMaskBits            = 7
     };
 
     unsigned Bits; // Naturally sized for speed.
 
-    ExtInfo(unsigned Bits) : Bits(Bits) {}
+  public:
+    struct Uncommon {
+      // We preserve a full clang::Type *, not a clang::FunctionType *, so
+      // we can keep sugar in case we need to present an error to the user.
+      const clang::Type *ClangFunctionType;
```

Comment:
This is what we want for the formal type `AnyFunctionType`, but it probably shouldn't be true of `SILFunctionType`, which is always canonical and therefore should probably be storing a canonical Clang type as well.  That should at least be explained here, since you're reusing the type in both places.

Reply:
> That should at least be explained here, since you're reusing the type in both places.

Ah, the name is the same in both places but the structure is different. Right now, I'm storing a `clang::FunctionType *` in `SILFunctionType`, but yeah, I should store a canonical type instead.

Reply:
We decided to use `clang::FunctionType *` to avoid a dependency on the clang's Type.h in Types.h. We will maintain the invariant that it is canonical by hand.